src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecutionList.java

```diff
@@ -252,4 +233,43 @@ public void onFailure(Throwable t) {
         executor.shutdown();
         executor.awaitTermination(1, TimeUnit.MINUTES);
     }
+
+    /**
+     * Whenever a Pipeline resumes, resume all incomplete steps in its {@link FlowExecution}.
+     *
+     * Called by {@code WorkflowRun.onLoad}, so guaranteed to run if a Pipeline resumes regardless of its presence in
+     * {@link FlowExecutionList}.
+     */
+    @Extension
+    public static class ResumeStepExecutionListener extends FlowExecutionListener {
+        @Override
+        public void onResumed(@Nonnull FlowExecution e) {
+            Futures.addCallback(e.getCurrentExecutions(false), new FutureCallback<List<StepExecution>>() {
+                @Override
+                public void onSuccess(List<StepExecution> result) {
+                    FlowExecutionList list = FlowExecutionList.get();
+                    FlowExecutionOwner owner = e.getOwner();
+                    if (!list.runningTasks.contains(owner)) {
```

Comment:
Maybe check `isComplete` first?

Reply:
Yeah I was thinking about that as well. `CpsFlowExecution.isComplete` is checked in `CpsFlowExecution.onLoad` before `loadProgramAsync` gets called, but `WorkflowRun.onLoad` only checks `CpsFlowExecution.isComplete()` if `WorkflowRun.completed` is `null`, which I think should only be true for old Pipelines from before `WorkflowRun.completed` was added (which seems wrong, probably better to always check in case there is a mismatch between the serialized state of the execution and the run, so I will look at that in a bit).

I also wasn't sure if it would be possible for there to be a timing issue, for example if the pipeline was suspended somewhere where it doesn't need any additional input to proceed and so the CPS VM thread finishes the build while we are on another thread still in `WorkflowRun.onLoad`.

Reply:
> if it would be possible for there to be a timing issue

Almost certainly.

Reply:
> Almost certainly.

If this is the case though, then we cannot fix the issue here, because there is no synchronization that can guarantee that the build does not complete right after we check `isComplete` but before we do anything else. Maybe the CPS VM thread should not attempt to execute anything until `WorkflowRun.onLoad` finishes or something like that.

Reply:
What makes you think the observed problem is the result of a race condition?

Reply:
I'm not following your question. I just mean that it is not clear to me whether checking `isComplete` here actually gives us any useful guarantees, partly because `WorkflowRun.onLoad` is not guaranteed to check it and partly because the execution actually completing will happen on the CPS VM thread, which is unrelated to the thread calling `WorkflowRun.onLoad`. I don't necessarily think that the underlying reported problem is a race condition.

Reply:
When I said

> then we cannot fix the issue here,

I just meant that we cannot fix any timing issues related to checking `isComplete` or not. Maybe that was not clear.

Reply:
Ah ok, so reading through `WorkflowRun.onLoad` again, it _is_ guaranteed to call `isComplete` inside of `WorkflowRun.getExecution` [here](https://github.com/jenkinsci/workflow-job-plugin/blob/1f016e0f8b4c812b63b40d6d15d70ffc20a94ea5/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L701-L704), and if the execution is complete, it marks the `WorkflowRun` as complete (this was the fix for the symptoms of [JENKINS-50199](https://issues.jenkins.io/browse/JENKINS-50199)), so [this condition](https://github.com/jenkinsci/workflow-job-plugin/blob/1f016e0f8b4c812b63b40d6d15d70ffc20a94ea5/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L565) would be false and so `FlowExecutionListener.fireResumed` would be skipped along with `ResumeStepExecutionListener.onResumed` if the execution was marked as complete in its serialized form.

That leaves us with a potential case where `WorkflowRun.getExecution` loads and resumes the execution, which has not yet completed, but then the execution completes before `ResumeStepExecutionListener.onResumed` has been called by `WorkflowRun.onLoad` after `WorkflowRun.getExecution` and so we resume the steps unnecessarily. We could check `isComplete` here in case this happens, but it seems somewhat unlikely, and the issue is even if we do check it, the execution could complete right after the check before `ResumeStepExecutionListener.onResumed` attempts to resume the step executions, so we wouldn't really be guaranteed to prevent the problem. We could try to modify execution resumption to somehow only start after `WorkflowRun.onLoad` returns, but that seems like a somewhat risky change unless we have some evidence that it is a problem in practice.

I think I'll go ahead and add the `isComplete` check since it won't hurt and could potentially help in some case, but I'll also add a comment mentioning that execution is asynchronous and so there could still be problems.

Reply:
Check added with a comment in e9f2421.